### ISSUE TITLE
Anomaly File "pretyping/cases.ml", line 2027, characters 2-8: Assertion failed." with let-in in arity of inductive type

### ISSUE THREAD

**Issue by akr** [coq/coq#11030]
#### Description of the problem

I found an anomaly.

```
% coqtop
Welcome to Coq 8.10.1 (November 2019)

Coq < Inductive I : let z := 0 in Set := C : I.
I is defined
I_rect is defined
I_ind is defined
I_rec is defined

Coq < Check match C with C => true end.
Toplevel input, characters 0-33:
> Check match C with C => true end.
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error:
Anomaly
"File "pretyping/cases.ml", line 2027, characters 2-8: Assertion failed."
Please report at http://coq.inria.fr/bugs/.
```

#### Coq Version

```
% coqtop -v
The Coq Proof Assistant, version 8.10.1 (November 2019)
compiled on Nov 2 2019 22:43:49 with OCaml 4.09.0
```

**Comment by Janno** [comment]
Worked in 8.9. `git bisect` points to:

```
29f749d87aebb8226bb9da624c57f83942881a99 is the first bad commit
commit 29f749d87aebb8226bb9da624c57f83942881a99
Author: Hugo Herbelin <Hugo.Herbelin@inria.fr>
Date:   Sat Aug 20 17:13:58 2016 +0200

    Trying an abstracting dependencies heuristic for the match return clause even when no type constraint is given.
    
    This no-inversion and maximal abstraction over dependencies in (rel)
    variables heuristic was used only when a type constraint was given.
    
    By doing so, we ensure that all three strategies "inversion with
    dependencies as evars", "no-inversion and maximal abstraction over
    dependencies in (rel) variables", "no-inversion and no abstraction
    over dependencies" are tried in all situations where a return clause
    has to be inferred.
    
    See penultimate commit for discussion.

 pretyping/cases.ml          | 25 +++++++++++--------------
 test-suite/success/Case13.v | 12 ++++++++++++
 2 files changed, 23 insertions(+), 14 deletions(-)
```

The corresponding PR coq/coq#262 took a long time to be merged into master, hence the 2016 date which really threw me off.

`git log 29f749d87aebb8226bb9da624c57f83942881a99` contains a lot of interesting comments that probably explain what happened but I don't understand them.

**Comment by herbelin** [comment]
Fixed in #18921. The commit 29f749d87ae introduced a detour via the function building a `return` predicate for `Program`'s `match` which was not supporting the presence of `let` in the arity of an inductive type.
